base_layer/core/src/base_node/states/horizon_state_sync/validators/headers.rs

```diff
@@ -0,0 +1,250 @@
+                target: LOG_TARGET,
+                "Recorded header target difficulty was incorrect: (got = {}, expected = {})",
+                block_header.pow.target_difficulty,
+                target
+            );
+            return Err(ValidationError::BlockHeaderError(
+                BlockHeaderValidationError::ProofOfWorkError(PowError::InvalidTargetDifficulty),
+            ));
+        }
+
+        let achieved = block_header.achieved_difficulty();
+        if achieved < target {
+            warn!(
+                target: LOG_TARGET,
+                "Proof of work for {} was below the target difficulty. Achieved: {}, Target:{}",
+                block_header.hash().to_hex(),
+                achieved,
+                target
+            );
+            return Err(ValidationError::BlockHeaderError(
+                BlockHeaderValidationError::ProofOfWorkError(PowError::AchievedDifficultyTooLow),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// This function tests that the block timestamp is greater than the median timestamp at the specified height.
+    pub fn check_median_timestamp(
+        &self,
+        block_header: &BlockHeader,
+        tip_header: &BlockHeader,
+    ) -> Result<(), ValidationError>
+    {
+        if self.is_genesis(block_header) {
+            // Median timestamps check not required for the genesis block header
+            return Ok(());
+        }
+
+        let start_height = block_header
+            .height
+            .saturating_sub(self.rules.consensus_constants().get_median_timestamp_count() as u64);
+
+        if start_height == tip_header.height {
+            return Ok(());
+        }
+
+        let block_nums = (start_height..tip_header.height).collect();
+        let mut timestamps = self
+            .db
+            .fetch_headers(block_nums)
+            .map_err(|e| ValidationError::CustomError(e.to_string()))?
+            .iter()
+            .map(|h| h.timestamp)
+            .collect::<Vec<_>>();
+        timestamps.push(tip_header.timestamp);
+
+        // TODO: get_median_timestamp incorrectly returns an Option
+        let median_timestamp =
+            get_median_timestamp(timestamps).expect("get_median_timestamp only returns None if `timestamps` is empty");
+
+        if block_header.timestamp < median_timestamp {
+            warn!(
+                target: LOG_TARGET,
+                "Block header timestamp {} is less than median timestamp: {} for block:{}",
+                block_header.timestamp,
+                median_timestamp,
+                block_header.hash().to_hex()
+            );
+            return Err(ValidationError::BlockHeaderError(
+                BlockHeaderValidationError::InvalidTimestamp,
+            ));
+        }
+        Ok(())
+    }
+
+    /// Returns the set of target difficulties for the specified proof of work algorithm.
+    fn fetch_target_difficulties(
+        &self,
+        block_header: &BlockHeader,
```

Comment:
should this not just be changed to pow::algo?
Thats all this function uses of the block header

Reply:
But looking at the above function, it looks like this call will fail if you dont test on tip.

Reply:
I think you're right, I wrote this a while ago so I'll look into this when I get time

Reply:
About passing in of PoW, it's debatable - it's a private implementation detail so I don't really have a strong opinion on that. The comment should be updated though. We're "fetching the target difficulty from the header" which makes just as much sense as "fetching the target difficulty from the pow algorithm" in the context of the block header validator so.. whatever ;) I'll change it to the latter.

Reply:
The block header used for logging purposes, so that tips it for me in favour of keeping it `&BlockHeader`